File: Leetcode/sbgo/Leetcode802-找到最终的安全状态.go
package sbgo

func eventualSafeNodes(graph [][]int) []int {
	n := len(graph)
	rev := make([][]int, n)
	d := make([]int, n)
	for i := 0; i < n; i++ {
		for _, b := range graph[i] {
			a := i
			rev[b] = append(rev[b], a)
			d[a]++
		}
	}
	que := []int{}
	for i := 0; i < n; i++ {
		if d[i] == 0 {
			que = append(que, i)
		}
	}
	for len(que) > 0 {
		t := que[0]
		que = que[1:]
		for _, b := range rev[t] {
			d[b]--
			if d[b] == 0 {
				que = append(que, b)
			}
		}
	}
	res := []int{}
	for i := 0; i < n; i++ {
		if d[i] == 0 {
			res = append(res, i)
		}
	}
	return res
}

File: Leetcode/PY/Leetcode802-找到最终的安全状态.py
from typing import List

class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        rev = [[] for _ in range(n)]
        d = [0] * n
        for i in range(n):
            for b in graph[i]:
                a = i
                rev[b].append(a)
                d[a] += 1
        que = []
        for i in range(n):
            if d[i] == 0:
                que.append(i)

        while len(que) > 0:
            t = que[0]
            que.pop(0)
            for b in rev[t]:
                d[b] -= 1
                if d[b] == 0:
                    que.append(b)
        res = []
        for i in range(n):
            if d[i] == 0:
                res.append(i)
        return res
            